### PR TITLE
feat: added authentication and private routes

### PR DESCRIPTION
**Pull request by SanjayJatti**
- Added Authentication feature
- Added Private Route
Files to Review:
- AuthContext.js
- AuthReducer.js
- PrivateRoute.js
- Signup.jsx
- Login.jsx
authentication

### CODE REVIEW

File: src/Components/PrivateRoute/PrivateRoute.jsx [hunk 1]
```diff
@@ -1,7 +1,14 @@
 import React from 'react'
+import { Navigate } from 'react-router-dom';
+import { useAuth } from '../../Context/AuthContext'
 
-export const PrivateRoute = () => {
+export const PrivateRoute = ({children}) => {
+  const { authState } = useAuth()
+  const {isUserLoggedIn } = authState;
+  console.log(isUserLoggedIn)
```
Comment (prathmeshjagtap) [inline src/Components/PrivateRoute/PrivateRoute.jsx]: remove console.log

File: src/Components/Header/Header.jsx [hunk 2]
```diff
@@ -27,11 +30,19 @@ const Header = () => {
...
+          {!isUserLoggedIn ? (
             <NavLink to="/login" className="nav-link">
               <button className="btn btn-primary">LogIn</button>
             </NavLink>
-          </li>
+          ) : (
+            <div className="nav-link">
+              <button 
+              className="btn btn-primary"
+              onClick={()=> logOutHandler()}>Logout</button>
```
Comment (prathmeshjagtap) [inline src/Components/Header/Header.jsx]: no need of arrow function here

File: src/Context/AuthContext.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  };
+
+  const logOutHandler = () => {
+    localStorage.removeItem("token");
+    authDispatch({
+      type: IS_USER_LOGGED_IN,
+    });
+    navigate("/");
+  };
+console.log(authState.isUserLoggedIn)
```
Comment (prathmeshjagtap) [inline src/Context/AuthContext.js]: console.log

**Review by prathmeshjagtap — APPROVED** [review]
rest everything is good and UI is also good

File: src/Context/AuthContext.js (hunk 3 above)
Comment (SanjayJatti) [reply src/Context/AuthContext.js]: Done 956201028c70f7e49077aa2b5493288e3226a945

File: src/Context/AuthContext.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    }
+  };
+
+  const logOutHandler = () => {
+    localStorage.removeItem("token");
+    authDispatch({
+      type: IS_USER_LOGGED_IN,
+    });
+    navigate("/");
+  };
```
Comment (Hardiegogo) [inline src/Context/AuthContext.js]: You can abstract these functions out and put them in a file under the name "auth-actions" or something.

**Review by Hardiegogo — COMMENTED** [review]
Good code, I have suggested a very small change above for better structure and readability.

File: src/Context/AuthContext.js (hunk 4 above)
Comment (SanjayJatti) [reply src/Context/AuthContext.js]: I have corrected it b25c3be573cac7ea01f469bd63c031a3a48a9496. Thanks for review!

**Pull request merged**